packages/core/src/RenderPipeline/PrimitiveChunk.ts

```diff
@@ -0,0 +1,186 @@
+
+    // Indices
+    const indexBuffer = new Buffer(engine, BufferBindFlag.IndexBuffer, maxVertexCount * 8, BufferUsage.Dynamic, true);
+    primitive.setIndexBufferBinding(new IndexBufferBinding(indexBuffer, IndexFormat.UInt16));
+
+    this.primitive = primitive;
+    this.vertices = new Float32Array(vertexBuffer.data.buffer);
+    this.indices = new Uint16Array(indexBuffer.data.buffer);
+    this.vertexFreeAreas.push(new Area(0, maxVertexCount * 9));
+  }
+
+  allocateSubChunk(vertexCount: number): SubPrimitiveChunk | null {
+    const area = this._allocateArea(this.vertexFreeAreas, vertexCount * 9);
+    if (area) {
+      const subChunk = this.subChunkPool.get();
+      subChunk.chunk = this;
+      subChunk.vertexArea = area;
+
+      const subMesh = this.subMeshPool.get();
+      subMesh.topology = MeshTopology.Triangles;
+      subChunk.subMesh = subMesh;
+      return subChunk;
+    }
+
+    return null;
+  }
+
+  freeSubChunk(subChunk: SubPrimitiveChunk): void {
+    this._freeArea(this.vertexFreeAreas, subChunk.vertexArea);
+    this.subMeshPool.return(subChunk.subMesh);
+    this.subChunkPool.return(subChunk);
+  }
+
+  uploadBuffer(): void {
+    // Set data option use Discard, or will resulted in performance slowdown when open antialias and cross-rendering of 3D and 2D elements.
+    // Device: iphone X(16.7.2)、iphone 15 pro max(17.1.1)、iphone XR(17.1.2) etc.
+    const { primitive, updateVertexStart, updateVertexLength } = this;
+    if (updateVertexStart !== Number.MAX_SAFE_INTEGER && updateVertexLength !== Number.MIN_SAFE_INTEGER) {
+      primitive.vertexBufferBindings[0].buffer.setData(
+        this.vertices,
+        updateVertexStart * 4,
+        updateVertexStart,
+        updateVertexLength,
+        SetDataOptions.Discard
+      );
+
+      this.updateVertexStart = Number.MAX_SAFE_INTEGER;
+      this.updateVertexLength = Number.MIN_SAFE_INTEGER;
+    }
+
+    primitive.indexBufferBinding.buffer.setData(this.indices, 0, 0, this.updateIndexLength, SetDataOptions.Discard);
+    this.updateIndexLength = 0;
+  }
+
+  destroy(): void {
+    this.primitive._addReferCount(-1);
+    this.primitive.destroy();
+    this.primitive = null;
+    this.vertices = null;
+    this.indices = null;
+    this.areaPool.garbageCollection();
+    this.areaPool = null;
+  }
+
+  private _allocateArea(entries: Area[], needSize: number): Area | null {
+    const pool = this.areaPool;
+    for (let i = 0, n = entries.length; i < n; ++i) {
+      const area = entries[i];
+      const size = area.size;
+      if (size > needSize) {
+        const newArea = pool.get();
+        newArea.start = area.start;
+        newArea.size = needSize;
+        area.start += needSize;
+        area.size -= needSize;
+        return newArea;
+      } else if (size === needSize) {
+        entries.splice(i, 1);
+        return area;
+      }
+    }
+    return null;
+  }
+
+  private _freeArea(areas: Area[], area: Area): void {
```

Comment:
private _freeArea(area: Area): void {
    const areas = this.vertexFreeAreas;
    ……
}